horovod/mxnet/__init__.py

```diff
@@ -100,18 +100,21 @@ def __init__(self, params, optimizer, optimizer_params=None):
     def _allreduce_grads(self):
         if size() == 1: return
 
+        # In MXNet 2.0, param.name is no longer unique.
+        # Meanwhile, since horovod requires Python 3.6, there is no need to sort
+        # self._params as enumerating a python dict is always deterministic.
         for i, param in enumerate(self._params):
             if param.grad_req != 'null':
                 allreduce_(param.list_grad()[0], average=False,
-                           name=param.name, priority=-i)
+                           name=str(i), priority=-i)
```

Comment:
Could you elaborate more on the `param.name` being no longer unique in MXNet 2.0? I'd hesitate to revert back to using `str(i)` as the allreduce request names, as it would reintroduce the problem described in #1679. Basically, in cases where users may have multiple `DistributedTrainer`s, this naming scheme will not differentiate between gradients being submitted by the different optimizers, instead producing multiple requests with the same name (i.e. multiple `allreduce.0`, `allreduce.1`, etc.).

I see in your changes to `broadcast_parameters` you use the dictionary key values, but I guess that isn't an option here because `self._params` is just a list of params, rather than the original dictionary. 

Maybe one option would be to add an optional `name` argument for the `DistributedTrainer` that gets prepended to the names of the allreduce operations being submitted? That way, users can provide unique trainer names if they need to disambiguate allreduce operations from multiple trainers.

Reply:
I'm adding a parameter -> name mapping in the constructor (`self._param2name`). Maybe we can use that, too.

Reply:
I think using `name = self._param2name[p._uuid]` after https://github.com/apache/incubator-mxnet/pull/18877 lands might be a cleaner option than adding a global name to the `DistributedTrainer`. Aligns more closely to the existing code and would be less disruptive to users who might be using multiple optimizers already. What do you think?

Reply:
cc @leezu to suggest on the choice of unique names for parameters.

Reply:
The names in `self._param2name[p._uuid]` will not be globally unique if the user creates multiple Blocks with the same structure but different parameters. Actually I'm not convinced we should add `self._param2name[p._uuid]`

Reply:
@romerojosh @eric-haibin-lin what's the uniqueness and consistency requirement for this identifier? Does it need to be consistent across different workers?

Reply:
The `name` of the allreduce operation needs to be consistent across all workers participating in the communication. This name identifier is what is used to match up tensors in the Horovod backend. Concerning uniqueness, names can be reused; however, all operations in flight must have unique names (i.e. a name can only be reused if the previous Horovod operation using that name has completed).

Reply:
In fact after checking with @leezu, the following two blocks will have the same name for parameters in mxnet2:
```
net1 = nn.Dense(10)
net2 = nn.Dense(10)

assert net1.collect_params().keys() == net2.collect_params().keys()
```

So for horovod we probably need to go back to the approach where we add a `name` argument to DistributedTrainer.